### PR TITLE
add support for rect mark

### PR DESCRIPTION
**Pull request by haldenl**
Support rect markings in search space.

### CODE REVIEW

File: src/constraint/spec.ts [hunk 1]
```diff
@@ -657,7 +659,15 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
             }
             return false;
           }
-
+        case Mark.RECT:
+          // Until CompassQL supports layering, it only makes sense for
+          // rect to encode DxD (otherwise just use bar).
+          const xEncQ = specM.getEncodingQueryByChannel(Channel.X);
+          const yEncQ = specM.getEncodingQueryByChannel(Channel.Y);
+          const xIsDimension = isDimension(xEncQ);
```
Comment (kanitw) [inline src/constraint/spec.ts]: It doesn't always have to be a dimension (1D) works too.

**Comment by haldenl** [comment]
I made a few changes:
- added a test for DxDxQ in generate.test.ts (not sure if this is the correct place to put it, recommend.test.ts looks similarly sparse).
- adding support for 1xD + color rect encoding without generating bad results required a special case in the 'omitNonPositionalOrFacetOverPositionalChannels' and adding another constraint 'omitFacetOverPositionalChannels'. They seem a little redundant together, I was wondering if it would make sense to split 'omitNonPositionalOrFacetOverPositionalChannels' into two separate tests, one for NonPositional and another for Facet.

File: src/ranking/effectiveness/mark.ts [hunk 2]
```diff
@@ -262,8 +262,8 @@ function init() {
     DISCRETE_OR_NONE.forEach((yType) => {
       // has occlusion
       const ddMark = {
-        point: 0,
         rect: 0,
+        point: -0.0025,
```
Comment (kanitw) [inline src/ranking/effectiveness/mark.ts]: I wonder if there are corner cases where making point inferior to rect in this case cause a bit more problem.

File: test/generate.test.ts [hunk 3]
```diff
@@ -299,6 +299,35 @@ describe('generate', function () {
...
+        },{
+          channel: SHORT_WILDCARD,
+          field: 'Q',
+          type: Type.QUANTITATIVE
+        }]
+      };
+
+      const answerSet = generate(specQ, schema, CONFIG_WITH_AUTO_ADD_COUNT);
+      it('should generate a rect heatmap with color encoding the quantitative measure', () => {
+        answerSet.forEach((specM) => {
```
Comment (kanitw) [inline test/generate.test.ts]: Currently the test checks if the result contains rect with color. 
We may want to write another test that make sure we recommend rect with color over other cases.  

(This test is good, but just we may want to add more test.)

File: test/ranking/effectiveness/mark.test.ts [hunk 4]
```diff
@@ -217,7 +217,7 @@ export const DD_RULESET = {
     const _rules: Rule[] = [];
     [TIMEUNIT_O, O, BIN_Q, N].forEach((xType) => {
       [TIMEUNIT_O, O, BIN_Q, N].forEach((yType) => {
-        const ddRank = [POINT, TEXTMARK, TICK, [BAR, LINE, AREA], RULE];
+        const ddRank = [RECT, POINT, TEXTMARK, TICK, [BAR, LINE, AREA], RULE];
```
Comment (kanitw) [inline test/ranking/effectiveness/mark.test.ts]: REMOVE the todo below about rect

File: src/constraint/spec.ts [hunk 5]
```diff
@@ -429,37 +430,64 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
-        } else if (channel === Channel.Y) {
-          hasY = true;
-        } else if (!isWildcard(channel)) {
-          // All non positional channel / Facet
-          hasNonPositionalChannelOrFacet = true;
-          if (specM.wildcardIndex.hasEncodingProperty(i, Property.CHANNEL)) {
-            hasEnumeratedNonPositionOrFacetChannel = true;
+      switch(mark) {
+        case Mark.RECT:
+          // RECT is a special exception, as it can be used in a 1xD manner
```
Comment (kanitw) [inline src/constraint/spec.ts]: I'm not sure if this constraint should be relaxed for rect?

If there is rect 1D with color, it is still better to use bar 2D (move color to another position channel) anyway?

File: src/constraint/spec.ts [hunk 6]
```diff
@@ -657,7 +685,26 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
+          // rect to encode DxD or 1xD (otherwise just use bar).
+          // Furthermore, color should only be used in a 'heatmap' fashion
+          // (with a measure field).
+          const xEncQ = specM.getEncodingQueryByChannel(Channel.X);
+          const yEncQ = specM.getEncodingQueryByChannel(Channel.Y);
+          const xIsDimension = isDimension(xEncQ);
+          const yIsDimension = isDimension(yEncQ);
+
+          const colorEncQ = specM.getEncodingQueryByChannel(Channel.COLOR);
+          const colorIsQuantitative = isMeasure(colorEncQ);
```
Comment (kanitw) [inline src/constraint/spec.ts]: What if color is ordinal?   (In the case where the data have no duplicate pairs of x, y values.)

File: src/constraint/spec.ts [hunk 7]
```diff
@@ -429,37 +430,64 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
+              (opt.constraintManuallySpecifiedValue && hasNonPositionalChannelOrFacet)
+            ) {
+            return hasX && hasY;
+          }
+          return true;
       }
+    }
+  },
+  {
+    name: 'omitFacetOverPositionalChannels',
```
Comment (kanitw) [inline src/constraint/spec.ts]: I still do not understand why you need to split this constraint from the other.  Could you please explain more?

File: test/generate.test.ts (hunk 3 above)
Comment (kanitw) [reply test/generate.test.ts]: Perhaps, we may need to add more test to the ranking part?

File: src/constraint/spec.ts (hunk 5 above)
Comment (haldenl) [reply src/constraint/spec.ts]: If I recall correctly, in one of our previous conversations you mentioned this 1D w/ color should be a possibility (and that's why I relaxed the constraint). I think it does make more sense to just use a bar in that case, however.

File: src/constraint/spec.ts (hunk 7 above)
Comment (haldenl) [reply src/constraint/spec.ts]: The split was motivated by 1D rect support. The original constraint also meant 'color' could not be used unless both positional channels had been filled (which would eliminate 1D rect), hence the exception for rect in that constraint. Without this second constraint, rows or columns appeared even for 1D rect encodings (e.g. multiple 1D rect encodings lined up). My thinking was that rect should be exempt from the first constraint, but all encodings should match the second constraint. It is a little redundant, hence my comment about splitting into just 'nonpositional' (maybe there is a better word for this that just encompasses color / size but excludes 'row' / 'column') and 'facet'.

File: src/constraint/spec.ts (hunk 6 above)
Comment (haldenl) [reply src/constraint/spec.ts]: done!

File: test/ranking/effectiveness/mark.test.ts (hunk 4 above)
Comment (haldenl) [reply test/ranking/effectiveness/mark.test.ts]: done

File: src/ranking/effectiveness/mark.ts (hunk 2 above)
Comment (haldenl) [reply src/ranking/effectiveness/mark.ts]: Since rect should only compete with point in a DxDx(Q|O) query, there should only be two cases right? (seen below).

screen shot 2017-11-19 at 4 07 09 pm
screen shot 2017-11-19 at 4 54 31 pm

File: test/generate.test.ts (hunk 3 above)
Comment (haldenl) [reply test/generate.test.ts]: Would these go in `mark.test.ts`? I see a Discrete-Discrete test there (where Rect > Point) but I'm not clear on how the tests in that file are able to handle 3rd dimensions like (with color Rect should be > Point but if the 3rd channel is unspecified then size is better and Point > Rect).

File: src/ranking/effectiveness/mark.ts (hunk 2 above)
Comment (haldenl) [reply src/ranking/effectiveness/mark.ts]: Ah another case where we specific color and therefore Rect should beat Point

image

**Comment by kanitw** [comment]
Not sure why there are so many new "result" files in this PR.  Could you please squash and rebase to the latest master to make this contain only relevant diff?  That will make it easier for me to review. (so I can focus on the core part of this PR.)

File: src/ranking/effectiveness/mark.ts (hunk 2 above)
Comment (kanitw) [reply src/ranking/effectiveness/mark.ts]: If there is x:D,y:D,?:Q where Q is not aggregated, then perhaps points are better because at least users can notice the occlusion. 

If there is ?:D,?:D,?:Q  where Q is not aggregated, perhaps, having one of the Q on x or y maybe better?

We should also consider the cases for Q is aggregated -- that case is actually less problematic.

**Comment by kanitw** [comment]
@haldenl  -- FYI, I've rebased this branch on top of https://github.com/vega/compassql/pull/422 and force pushed.   Make sure to remove your local branch + checkout the new one before making further changes.

File: src/constraint/spec.ts (hunk 5 above)
Comment (kanitw) [reply src/constraint/spec.ts]: >  I think it does make more sense to just use a bar in that case, however.

I see.  I agree that bar still makes more sense anyway. 
Since this is a non-strict constraint, maybe we don't have to create a special case for rect?


In a way, point with x and size/color is a fair game too.  It's just that scatterplot is almost always more reasonable.  So it seems a bit inconsistent to relax this constraint only for rect.

Comment (haldenl) [reply src/constraint/spec.ts]: That sounds fair, I will remove this case

**Comment by haldenl** [comment]
I don't believe it makes sense to split the constraint if we do not want to relax the first constraint for 1D rect. I will remove the second constraint!

File: src/ranking/effectiveness/mark.ts (hunk 2 above)
Comment (haldenl) [reply src/ranking/effectiveness/mark.ts]: For now assume occlusion heuristics are correct and modify rankings accordingly

File: src/config.ts [hunk 8]
```diff
@@ -45,6 +45,7 @@ export interface QueryConfig {
   omitRawWithXYBothOrdinalScaleOrBin?: boolean;
   omitRepeatedField?: boolean;
   omitNonPositionalOrFacetOverPositionalChannels?: boolean;
+  omitFacetOverPositionalChannels?: boolean;
```
Comment (kanitw) [inline src/config.ts]: Do we still need to add this config?

File: src/ranking/effectiveness/mark.ts (hunk 2 above)
Comment (haldenl) [reply src/ranking/effectiveness/mark.ts]: DxD Point + size >  DxD Rect + color > DxD Point + color

- Add a feature color channel to correct this.

- Because point > rect and point probably get 0; rect will get small minus (-R).  the rect x color can be a +R  and maybe do something small with point and color (minus  < |R|)

**Comment by haldenl** [comment]
This should be good to go. To summarize the example changes, we see RECT used whenever we see POINT used for Dx1, DxD, or DxDxQ, with arbitrary ordering between the two. Unless the query specifies RECT mark, the highest ranked recommendation based on effectiveness will be POINT + SIZE over RECT + COLOR.

File: test/ranking/effectiveness/mark.test.ts [hunk 9]
```diff
@@ -217,20 +217,19 @@ export const DD_RULESET = {
     const _rules: Rule[] = [];
     [TIMEUNIT_O, O, BIN_Q, N].forEach((xType) => {
       [TIMEUNIT_O, O, BIN_Q, N].forEach((yType) => {
-        const ddRank = [POINT, TEXTMARK, TICK, [BAR, LINE, AREA], RULE];
-
-        // TODO: RECT is not bad here.
-
+        const base = [TEXTMARK, TICK, [BAR, LINE, AREA], RULE];
```
[the remark this answers is not in the document]
Comment (kanitw) [reply test/ranking/effectiveness/mark.test.ts]: This is outdated ....

**Comment by haldenl** [comment]
@kanitw do you have any ideas as to why the travis builds are failing? 

>Failed to install 5. Remote repository may not be reachable.

**Comment by kanitw** [comment]
Not sure, but we can restart the build just in case it's due to server down.
